### PR TITLE
MULE-11334: Migrate old MuleClient uses

### PR DESCRIPTION
**Pull request by pablokraan**
_ Adding TestHttpClient rule to provide access to an HttpClient that uses the Http service
_ Migrating CXF test to use TestHttpClient instead of muleClient

### CODE REVIEW

File: compatibility/modules/cxf/src/test/java/org/mule/compatibility/module/cxf/CxfBackToBlockingTestCase.java [hunk 1]
```diff
@@ -59,26 +58,29 @@ protected void doSetUp() throws Exception {
...
-    InternalMessage result = client.send("http://localhost:" + dynamicPort.getNumber() + "/services/Echo",
-                                         InternalMessage.builder().payload(xml).mediaType(APP_SOAP_XML).build(),
-                                         HTTP_REQUEST_OPTIONS)
-        .getRight();
-    assertTrue(getPayloadAsString(result).contains("Hello!"));
-    String ct = result.getPayload().getDataType().getMediaType().toRfcString();
-    assertEquals("text/xml; charset=UTF-8", ct);
+
+    ParameterMap headersMap = new ParameterMap();
+    headersMap.put("content-type", APP_SOAP_XML.toRfcString());
```
Comment (elrodro83) [inline compatibility/modules/cxf/src/test/java/org/mule/compatibility/module/cxf/CxfBackToBlockingTestCase.java]: constant

File: compatibility/modules/cxf/src/test/java/org/mule/compatibility/module/cxf/CxfBackToBlockingTestCase.java [hunk 2]
```diff
@@ -59,26 +58,29 @@ protected void doSetUp() throws Exception {
...
+    HttpRequest request = HttpRequest.builder().setUri("http://localhost:" + dynamicPort.getNumber() + "/services/Echo")
+        .setMethod(POST.name()).setEntity(new InputStreamHttpEntity(xml)).setHeaders(headersMap).build();
+
+    HttpResponse response = httpClient.send(request, RECEIVE_TIMEOUT, false, null);
+    String payload = IOUtils.toString(((InputStreamHttpEntity) response.getEntity()).getInputStream());
+    assertTrue(payload.contains("Hello!"));
+    assertEquals("text/xml; charset=UTF-8", response.getHeaderValue("content-type"));
```
Comment (elrodro83) [inline compatibility/modules/cxf/src/test/java/org/mule/compatibility/module/cxf/CxfBackToBlockingTestCase.java]: constant

File: compatibility/modules/cxf/src/test/java/org/mule/compatibility/module/cxf/CxfBadSoapRequestTestCase.java [hunk 3]
```diff
@@ -39,24 +39,21 @@ protected String getConfigFile() {
...
+    HttpRequest request = HttpRequest.builder().setUri("http://localhost:" + dynamicPort.getNumber() + "/services/TestComponent")
+        .setMethod(POST.name()).setEntity(new ByteArrayHttpEntity(soapRequest.getBytes())).build();
 
-    assertNotNull(reply);
-    assertNotNull(reply.getPayload().getValue());
+    HttpResponse response = httpClient.send(request, RECEIVE_TIMEOUT, false, null);
 
-    String ct = reply.getPayload().getDataType().getMediaType().toRfcString();
-    assertEquals("text/xml; charset=UTF-8", ct);
+    assertEquals("text/xml; charset=UTF-8", response.getHeaderValue("content-type"));
```
Comment (elrodro83) [inline compatibility/modules/cxf/src/test/java/org/mule/compatibility/module/cxf/CxfBadSoapRequestTestCase.java]: constant

File: compatibility/modules/cxf/src/test/java/org/mule/compatibility/module/cxf/CxfBasicTestCase.java [hunk 4]
```diff
@@ -58,24 +57,27 @@ protected void doSetUp() throws Exception {
...
-                                         InternalMessage.builder().payload(xml).mediaType(APP_SOAP_XML).build(),
-                                         HTTP_REQUEST_OPTIONS)
-        .getRight();
-    assertTrue(getPayloadAsString(result).contains("Hello!"));
-    String ct = result.getPayload().getDataType().getMediaType().toRfcString();
-    assertEquals("text/xml; charset=UTF-8", ct);
+
+    HttpRequest request =
+        HttpRequest.builder().setUri("http://localhost:" + dynamicPort.getNumber() + "/services/Echo").setMethod(POST.name())
+            .setEntity(new InputStreamHttpEntity(xml)).addHeader("content-type", APP_SOAP_XML.toRfcString()).build();
```
Comment (elrodro83) [inline compatibility/modules/cxf/src/test/java/org/mule/compatibility/module/cxf/CxfBasicTestCase.java]: constant

File: compatibility/modules/cxf/src/test/java/org/mule/compatibility/module/cxf/CxfBasicTestCase.java [hunk 5]
```diff
@@ -58,24 +57,27 @@ protected void doSetUp() throws Exception {
...
+
+    HttpResponse response = httpClient.send(request, RECEIVE_TIMEOUT, false, null);
+
+    String payload = IOUtils.toString(((InputStreamHttpEntity) response.getEntity()).getInputStream());
+
+    assertTrue(payload.contains("Hello!"));
+    assertEquals("text/xml; charset=UTF-8", response.getHeaderValue("content-type"));
```
Comment (elrodro83) [inline compatibility/modules/cxf/src/test/java/org/mule/compatibility/module/cxf/CxfBasicTestCase.java]: constant

File: compatibility/modules/cxf/src/test/java/org/mule/compatibility/module/cxf/GZIPEncodingTestCase.java [hunk 6]
```diff
@@ -67,27 +65,36 @@ protected String getConfigFile() {
...
     validateResponse(response);
   }
 
-  private void validateResponse(InternalMessage response) throws Exception {
-    String unzipped = unzip(new ByteArrayInputStream(getPayloadAsBytes(response)));
+  private void validateResponse(HttpResponse response) throws Exception {
+    String unzipped = unzip(((InputStreamHttpEntity) response.getEntity()).getInputStream());
     assertThat(unzipped, compareXML(getAllResponse, unzipped).identical(), is(true));
-    assertThat(response.getInboundProperty(CONTENT_ENCODING), is(GZIP));
+    assertThat(response.getHeaderValue("content-encoding"), is(GZIP));
```
Comment (elrodro83) [inline compatibility/modules/cxf/src/test/java/org/mule/compatibility/module/cxf/GZIPEncodingTestCase.java]: constant

File: compatibility/modules/cxf/src/test/java/org/mule/compatibility/module/cxf/functional/CxfContentTypeNonBlockingTestCase.java [hunk 7]
```diff
@@ -39,39 +42,43 @@ protected String getConfigFile() {
...
-        .getRight();
-    String contentType = received.getInboundProperty("contentType");
+    HttpRequest httpRequest =
+        HttpRequest.builder().setUri("http://localhost:" + dynamicPort.getNumber() + "/helloClientProxy")
+            .setEntity(new ByteArrayHttpEntity("hello".getBytes()))
+            .setMethod(POST.name()).build();
+
+    HttpResponse httpResponse = httpClient.send(httpRequest, RECEIVE_TIMEOUT, false, null);
+
+    String contentType = httpResponse.getHeaderValue("content-type");
```
Comment (elrodro83) [inline compatibility/modules/cxf/src/test/java/org/mule/compatibility/module/cxf/functional/CxfContentTypeNonBlockingTestCase.java]: constant

File: compatibility/modules/cxf/src/test/java/org/mule/compatibility/module/cxf/functional/CxfContentTypeTestCase.java [hunk 8]
```diff
@@ -28,20 +29,24 @@
...
-    String contentType = received.getPayload().getDataType().getMediaType().toRfcString();
-    assertNotNull(contentType);
+    HttpRequest httpRequest =
+        HttpRequest.builder().setUri("http://localhost:" + dynamicPort.getNumber() + "/hello")
+            .setEntity(new ByteArrayHttpEntity(requestPayload.getBytes()))
+            .setMethod(POST.name()).build();
+
+    HttpResponse httpResponse = httpClient.send(httpRequest, RECEIVE_TIMEOUT, false, null);
+
+    String contentType = httpResponse.getHeaderValue("content-type");
```
Comment (elrodro83) [inline compatibility/modules/cxf/src/test/java/org/mule/compatibility/module/cxf/functional/CxfContentTypeTestCase.java]: constant

File: services/http/src/test/java/org/mule/services/http/TestHttpClient.java [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+import org.mule.service.http.api.domain.message.response.HttpResponse;
+import org.mule.services.http.impl.service.HttpServiceImplementation;
+
+import java.io.IOException;
+import java.util.concurrent.TimeoutException;
+
+import org.junit.rules.ExternalResource;
+
+/**
+ * Defines a {@link org.mule.service.http.api.client.HttpClient} using a defualt implementation of {@link HttpService}
```
Comment (elrodro83) [inline services/http/src/test/java/org/mule/services/http/TestHttpClient.java]: defualt -> default

File: tests/integration/src/test/java/org/mule/test/core/context/notification/ConnectorMessageErrorNotificationTestCase.java [hunk 10]
```diff
@@ -24,32 +21,32 @@
...
+    httpClient.send(request, TIMEOUT, false, null);
 
     assertNotifications();
   }
 
   @Override
   public RestrictedNode getSpecification() {
-    return new Node().parallel(new Node(ConnectorMessageNotification.class, MESSAGE_REQUEST_BEGIN, MULE_CLIENT_ID))
-        .parallel(new Node(ConnectorMessageNotification.class, MESSAGE_REQUEST_END, MULE_CLIENT_ID))
+    return new Node()
```
Comment (elrodro83) [inline tests/integration/src/test/java/org/mule/test/core/context/notification/ConnectorMessageErrorNotificationTestCase.java]: there notifications were just generated by the client?

File: tests/integration/src/test/java/org/mule/test/integration/domain/http/HttpSharePortTestCase.java [hunk 11]
```diff
@@ -46,25 +52,26 @@ protected String getDomainConfig() {
...
+    assertThat(payload, is("hello mule"));
   }
 
   protected SystemProperty getEndpointSchemeSystemProperty() {
     return new SystemProperty("scheme", "http");
   }
 
-  protected HttpRequestOptionsBuilder getOptionsBuilder() {
-    return newOptions();
+  protected TlsContextFactory getTlsContextFactory() {
```
Comment (elrodro83) [inline tests/integration/src/test/java/org/mule/test/integration/domain/http/HttpSharePortTestCase.java]: is it needed for this to return null or can this just be removed? (and the call on the builder too)

File: modules/service/src/main/java/org/mule/runtime/module/service/MuleServiceManager.java [hunk 12]
```diff
@@ -75,7 +75,13 @@ public void start() throws MuleException {
   private void startServices() throws MuleException {
     for (Service service : registeredServices) {
       if (service instanceof Startable) {
-        ((Startable) service).start();
+        ClassLoader originalContextClassLoader = Thread.currentThread().getContextClassLoader();
```
Comment (elrodro83) [inline modules/service/src/main/java/org/mule/runtime/module/service/MuleServiceManager.java]: static imports

File: compatibility/modules/cxf/src/test/java/org/mule/compatibility/module/cxf/CxfBadSoapRequestTestCase.java [hunk 13]
```diff
@@ -39,24 +40,20 @@ protected String getConfigFile() {
...
+    HttpRequest request = HttpRequest.builder().setUri("http://localhost:" + dynamicPort.getNumber() + "/services/TestComponent")
+        .setMethod(POST.name()).setEntity(new ByteArrayHttpEntity(soapRequest.getBytes())).build();
 
-    assertNotNull(reply);
-    assertNotNull(reply.getPayload().getValue());
+    HttpResponse response = httpClient.send(request, RECEIVE_TIMEOUT, false, null);
 
-    String ct = reply.getPayload().getDataType().getMediaType().toRfcString();
-    assertEquals("text/xml; charset=UTF-8", ct);
+    assertEquals("text/xml; charset=UTF-8", response.getHeaderValueIgnoreCase(CONTENT_TYPE));
```
Comment (elrodro83) [inline compatibility/modules/cxf/src/test/java/org/mule/compatibility/module/cxf/CxfBadSoapRequestTestCase.java]: use MediaType.XML.withCharset()

File: compatibility/modules/cxf/src/test/java/org/mule/compatibility/module/cxf/CxfBackToBlockingTestCase.java [hunk 14]
```diff
@@ -59,26 +59,29 @@ protected void doSetUp() throws Exception {
...
+
+    ParameterMap headersMap = new ParameterMap();
+    headersMap.put(CONTENT_TYPE, APP_SOAP_XML.toRfcString());
+    HttpRequest request = HttpRequest.builder().setUri("http://localhost:" + dynamicPort.getNumber() + "/services/Echo")
+        .setMethod(POST.name()).setEntity(new InputStreamHttpEntity(xml)).setHeaders(headersMap).build();
+
+    HttpResponse response = httpClient.send(request, RECEIVE_TIMEOUT, false, null);
+    String payload = IOUtils.toString(((InputStreamHttpEntity) response.getEntity()).getInputStream());
+    assertTrue(payload.contains("Hello!"));
+    assertEquals("text/xml; charset=UTF-8", response.getHeaderValueIgnoreCase(CONTENT_TYPE));
```
Comment (elrodro83) [inline compatibility/modules/cxf/src/test/java/org/mule/compatibility/module/cxf/CxfBackToBlockingTestCase.java]: use MediaType.XML.withCharset()

File: tests/integration/src/test/java/org/mule/test/core/context/notification/ConnectorMessageErrorNotificationTestCase.java (hunk 10 above)
Comment (pablokraan) [inline tests/integration/src/test/java/org/mule/test/core/context/notification/ConnectorMessageErrorNotificationTestCase.java]: Notification were genereated for mule client and flow. I removed the client ones

File: tests/integration/src/test/java/org/mule/test/integration/domain/http/HttpSharePortTestCase.java (hunk 11 above)
Comment (pablokraan) [inline tests/integration/src/test/java/org/mule/test/integration/domain/http/HttpSharePortTestCase.java]: Is used on the constructor, this test class does not uses a tlsContextFactory, but the subclass does

**Pull request merged**